k8s/kubicle/create.go

```diff
@@ -0,0 +1,644 @@
+	if err != nil {
+		return nil, err
+	}
+	pc.noProxy = os.Getenv("no_proxy")
+	return pc, nil
+}
+
+// TODO: Code copied from networking
+
+func validPhysicalLink(link netlink.Link) bool {
+	phyDevice := true
+
+	switch link.Type() {
+	case "device":
+	case "bond":
+	case "vlan":
+	case "macvlan":
+	case "bridge":
+		if strings.HasPrefix(link.Attrs().Name, "docker") ||
+			strings.HasPrefix(link.Attrs().Name, "virbr") {
+			phyDevice = false
+		}
+	default:
+		phyDevice = false
+	}
+
+	if (link.Attrs().Flags & net.FlagLoopback) != 0 {
+		return false
+	}
+
+	return phyDevice
+}
+
+func getFirstPhyDevice() string {
+	links, err := netlink.LinkList()
+	if err != nil {
+		return ""
+	}
+
+	for _, link := range links {
+		if !validPhysicalLink(link) {
+			continue
+		}
+
+		addrs, err := netlink.AddrList(link, netlink.FAMILY_V4)
+		if err != nil || len(addrs) == 0 {
+			continue
+		}
+
+		return addrs[0].IP.String()
+	}
+
+	return ""
+}
+
+func genToken() (string, error) {
+	var buf [11]byte
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	_, err := io.ReadFull(r, buf[:])
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s.%s",
+		hex.EncodeToString(buf[:3]), hex.EncodeToString(buf[3:])), nil
+}
+
+func (c *creator) createMasterConfig() {
+	mc := &masterConfig{
+		baseConfig: baseConfig{
+			VCPUs:        c.opts.masterVM.vCPUs,
+			RAM:          c.opts.masterVM.mem,
+			Disk:         c.opts.masterVM.disk,
+			User:         c.opts.user,
+			ImageUUID:    c.opts.imageUUID,
+			PublicKey:    c.pk,
+			Token:        c.token,
+			UserDataFile: "k8s-master-ci.yaml",
+			Description:  masterWorkloadName,
+		},
+		ExternalIP: c.opts.externalIP,
+	}
+
+	if mc.ExternalIP != "" {
+		mc.PhoneHomeIP = getFirstPhyDevice()
+	}
+
+	mc.HTTPSProxy = c.pc.httpsProxy
+	mc.HTTPProxy = c.pc.httpProxy
+	mc.NoProxy = c.pc.noProxy
+	c.mc = mc
+}
+
+func (c *creator) createWorkerConfig() {
+	wc := &workerConfig{
+		baseConfig: baseConfig{
+			VCPUs:        c.opts.masterVM.vCPUs,
+			RAM:          c.opts.masterVM.mem,
+			Disk:         c.opts.masterVM.disk,
+			User:         c.opts.user,
+			ImageUUID:    c.opts.imageUUID,
+			PublicKey:    c.pk,
+			Token:        c.token,
+			UserDataFile: "k8s-worker-ci.yaml",
+			Description:  workerWorkloadName,
+		},
+		MasterIP: c.masterIP,
+	}
+
+	wc.HTTPSProxy = c.pc.httpsProxy
+	wc.HTTPProxy = c.pc.httpProxy
+	wc.NoProxy = c.pc.noProxy
+	c.wc = wc
+}
+
+func (c *creator) createMasterWorkloadDefinition() error {
+	var buf bytes.Buffer
+
+	err := workloadTmpl.Execute(&buf, c.mc)
+	if err != nil {
+		return err
+	}
+
+	// TODO might want to check to see if these files already
+	// exist and error if a -f is not provided.
+
+	c.masterPath = filepath.Join(c.cwd, "k8s-master.yaml")
+	err = ioutil.WriteFile(c.masterPath, buf.Bytes(), 0600)
+	if err != nil {
+		return err
+	}
+
+	buf.Reset()
+
+	err = masterUDTmpl.Execute(&buf, c.mc)
+	if err != nil {
+		_ = os.Remove(c.masterPath)
```

Comment:
The goal is to create these two workload files atomically as c.masterPath references c.mc.UserDataFile.  It is useless without it so we delete it even if keep is specified.